### PR TITLE
refactor: download multiple versions of docs

### PR DESCRIPTION
**Pull request by HashimotoYT**
This PR makes the first step of changes for #288, this adds to collect script downloading multiple versions of Electron.

## What's versions we download now?

We download 4 stable versions of Electron where 1 of these it's `current` stable release what also contain other files like `electron-api.json` or website content. 3 others it's release branches in the `e/e`, please notice we don't get full version how we do for current, we only collect all data from the branch, so if change merged into this branch this gets fetch anyway if it's not published to npm. It's not bad, and not good but it simplifies the process.

## What's if release goes to EOL?

Rainbow dash takes this version and moves into Equestria 🦄. If this version goes to EOL this deletes. Easy peasy lemon squeezy.

Here not all easy how described above because when creates release branches with the number `10.0` we get an update already if this version doesn't publish to npm. So we move `6.0` goes to End Of Life fastly than the `10.0` released. Not sure if this very breaking change, because `10.0` should publish after some days after `9.0` goes stable.

## Files what need to review

- collect script: https://github.com/electron/i18n/blob/multi-ver-docs/script/collect.ts
- package.json: https://github.com/electron/i18n/blob/multi-ver-docs/package.json

All other changes it makes backward compatibility fixes for current docs.

## Automatically generate crowdin.yml

From now we generating crowdin.yml automatically when new release published. Regenerating starts only when folders more than `supportedVersions`. This file needs to fully know what's files need to translate. (not sure, but this makes from current, so I think this need to know what files need to upload).

If talks about this more, I'm not really sure if Crowdin deletes the previous folder structure, if not these not very big problems.

### Current translations

I'm really don't know what to be with current translations, for now, I'm manually moving current into own `current` folder, what's next, not sure. It's a task that need to exact before merging this.

### Folder structure

This changes folder structure, you can see the diff between the old and new below. From now we have 4 folders where 1 is a current stable release and 3 others it's released branches in `electron/electron`.

Previous folder structure:

```shell
$ tree content -d -L 2

├───en-US
│	└───docs
├───ru-RU
│   └───docs
```

New folder structure:

```shell
$ tree content -d -L 2

├───6-2-x
│   └───en-US
├───7-2-x
│   └───en-US
├───9-x-y
│   └───en-US
└───current
	└───en-US
```

You can see that some folders named `9-x-y`, some `7-2-x`, it's no problem for the current process since the current folder named in the new style.

Files like `website/locale.yml` and `website/blog/*.md` still exist in the current folder to prevent some breaking changes, also `electron-api.json` only exists in the current folder.

---

TODO list:

- [ ] Figure what's be with current translations.
- [ ] Should be website folder extracted from the current folder

Some small PS:

- Described above maybe describes not all that I'm want so if you have questions go.
- And yes for now for me this has more questions than answers.
- One big question, it's how to move the current translation.
- Please do not merge this until I am ready.

---

#### Edit 4/2/2020

Reading after 6 months it's fun. All code was rewritten to the current state. For now that not breaking CI, after the reviewing, I start breaking those things.

### CODE REVIEW

**Comment by BinaryMuse** [comment]
@HashimotoYT I'm looking over this PR now (on the command-line to simplify things).

One immediate question: do you know what will happen to the current translations on Crowdin when we merge this? Will they all be invalidated? Or are things smart enough to keep track?

**Comment by HashimotoYT** [comment]
> One immediate question: do you know what will happen to the current translations on Crowdin when we merge this? Will they all be invalidated? Or are things smart enough to keep track?

I don't know what be with the current translation. 😢

**Comment by BinaryMuse** [comment]
> > One immediate question: do you know what will happen to the current translations on Crowdin when we merge this? Will they all be invalidated? Or are things smart enough to keep track?
> 
> I don't know what be with the current translation. 😢

Okay. I will do some reading and reach out to Crowdin if I need to to figure this out. I would hate to lose all our current translations.

---

To simplify reviewing this PR, would you mind backing out the changes from the `content/` directory for now? The diff is just too big for GitHub to deal with. This should do the trick:

```bash
git reset origin/master -- content && git checkout -- content
```

This will leave the new folders under `content/` but will leave them out of the commit.

**Comment by HashimotoYT** [comment]
> To simplify reviewing this PR, would you mind backing out the changes from the content/ directory for now?

Looks like I'm doing something wrong, simplify to 100 changed files (maybe need merge from master), failing test should pass against when it's back to the previous structure.

**Comment by zeke** [comment]
cc @crowdin-support @andrulko there's some concern that this directory structure change could cause some translations to be lost. Can you confirm that our translation memory should protect us from that?

**Comment by Andrulko** [comment]
Hey @zeke! There is indeed data loss risk. You have TM as backup of course, but there is a way to keep all translations.

If you can spend time adjusting folders hierarchy in current `master` uploaded to Crowdin according to the hierarchy of refactored branch, then it will be possible to keep not only existing translations but also translations history (authorship).

Let's do a call next Tuesday to discuss details? I will do everything possible to make this migration easier!

File: lib/generate-crowdin-config.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { promises as fs } from 'fs'
+import * as path from 'path'
+import * as hbs from 'handlebars'
+
+export async function generateCrowdinConfig(versions: Array<string>) {
+  const template = await fs.readFile(path.join(__dirname, './templates/crowdin.yml.tpl'), { encoding: 'utf8'})
+  const generated = hbs.compile(template)({ versions })
+
+  fs.writeFile(path.join(__dirname, '../crowdin.yml'), generated)
```
Comment (malept) [inline lib/generate-crowdin-config.ts:9]:
```suggestion
  await fs.writeFile(path.join(__dirname, '../crowdin.yml'), generated)
```

File: lib/templates/crowdin.yml.tpl [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    translation: /content/%locale%/docs/tutorial/%original_file_name%
+  - source: /content/en-US/website/*.yml
+    translation: /content/%locale%/website/%original_file_name%
+  - source: /content/en-US/docs/api/*.md
+    translation: /content/%locale%/docs/api/%original_file_name%
+  - source: /content/en-US/docs/api/structures/*.md
+    translation: /content/%locale%/docs/api/structures/%original_file_name%
+  - source: /content/en-US/website/blog/*.md
+    translation: /content/%locale%/website/blog/%original_file_name%
+
```
Comment (malept) [inline lib/templates/crowdin.yml.tpl:20]: Why not always add `current` to `versions` when calling `generateCrowdinConfig`, so this isn't needed? It'll make it less error-prone to edit this file.

```suggestion
```

File: script/collect.ts [hunk 3]
```diff
@@ -12,8 +12,19 @@ import * as path from 'path'
 import { execSync } from 'child_process'
 import { Octokit } from '@octokit/rest'
 import { roggy, IResponse as IRoggyResponse } from 'roggy'
-const electronDocs = require('electron-docs')
-const englishBasepath = path.join(__dirname, '..', 'content', 'en-US')
+import { generateCrowdinConfig } from '../lib/generate-crowdin-config'
+import * as packageJson from '../package.json'
+const currentEnglishBasepath = path.join(
```
Comment (malept) [inline script/collect.ts:17]:
```suggestion
const currentEnglishBasePath = path.join(
```

File: script/collect.ts [hunk 4]
```diff
@@ -12,8 +12,19 @@ import * as path from 'path'
...
+  __dirname,
+  '..',
+  'content',
+  'current',
+  'en-US'
+)
+const englishBasepath = (version: string) =>
```
Comment (malept) [inline script/collect.ts:24]:
```suggestion
const englishBasePath = (version: string) =>
```

File: script/collect.ts [hunk 5]
```diff
@@ -64,15 +73,90 @@ async function fetchRelease() {
...
+    versions.pop()
+    await generateCrowdinConfig(versions)
+  }
+}
+
+async function delContent(branches: Array<string>) {
+  console.log('Deleting content:')
+
+  console.log('  - Deleting current content')
+  await del(currentEnglishBasepath)
```
Comment (malept) [inline script/collect.ts:121]: Future refactor: rename `del`

File: script/collect.ts [hunk 6]
```diff
@@ -64,15 +73,90 @@ async function fetchRelease() {
...
+  if (folders.length !== versions.length) {
+    versions.push('current')
+    const difference = folders.filter((x) => !versions.includes(x)).toString()
+    del(path.join(__dirname, '..', 'content', difference))
+    versions.pop()
+    await generateCrowdinConfig(versions)
+  }
+}
+
+async function delContent(branches: Array<string>) {
```
Comment (malept) [inline script/collect.ts:117]:
```suggestion
async function deleteContent(branches: Array<string>) {
```

File: script/collect.ts [hunk 7]
```diff
@@ -64,15 +73,90 @@ async function fetchRelease() {
...
+  const filteredBranches = Object.values(filtered)
+    .sort()
+    .slice(-NUM_SUPPORTED_VERSIONS)
+    .filter((arr) => arr !== currentVersion && arr !== 'current')
+
+  writeToPackageJSON('supportedVersions', filteredBranches)
+  console.log('Successfully written `supportedVersions` into package.json')
+}
+
+async function delUnsupportedBranches(versions: Array<string>) {
```
Comment (malept) [inline script/collect.ts:105]:
```suggestion
async function deleteUnsupportedBranches(versions: Array<string>) {
```

File: script/collect.ts [hunk 8]
```diff
@@ -64,15 +73,90 @@ async function fetchRelease() {
...
+  const docs = await roggy(release.tag_name, {
+    owner: 'electron',
+    repository: 'electron',
+  })
 
   docs
-    .filter((doc: IElectronDocsResponse) => doc.filename.startsWith('api/'))
-    .forEach(writeDoc)
+    .filter((doc) => doc.filename.startsWith('api/'))
+    .forEach((doc) => writeDoc(doc))
```
Comment (malept) [inline script/collect.ts:139]: Future refactor: figure out how to consolidate the 4 similar-looking blocks

File: script/collect.ts [hunk 9]
```diff
@@ -116,13 +203,38 @@ async function getMasterBranchCommit() {
...
+
+  return Promise.resolve()
+}
+
+async function fetchTutorialsFromSupportedBranch() {
+  console.log(`Feching tutorial docs from supported branches`)
+
+  for (const version of packageJson.supportedVersions) {
+    console.log(`  - from electron/electro#${version}`)
```
Comment (malept) [inline script/collect.ts:224]:
```suggestion
    console.log(`  - from electron/electron#${version}`)
```

**Comment by HashimotoYT** [comment]
🙋‍♂️ cc @crowdin-support @Andrulko

We started the finishing line of this PR and we want to ask some questions about making the migration and refactoring more safer and easier:

#### What steps needs from ours to implement this message:

> If you can spend time adjusting folders hierarchy in current master uploaded to Crowdin according to the hierarchy of refactored branch, then it will be possible to keep not only existing translations but also translations history (authorship).

(ref https://github.com/electron/i18n/pull/675#issuecomment-524379943 and https://github.com/electron/i18n/pull/675#issuecomment-524382735).

#### Also, I want to ask a question about the crowdin.yml file:

From now that file is generated automatically and hierarchy changes dynamically from version to version. The previous folders from  Crowdin panel are deleted automatically or we need to do some more specific for that.

**Comment by Andrulko** [comment]
Hi @HashimotoYT!

Currently, duplicates are visible and auto-translated inside [electron](https://crowdin.com/project/electron) project. It means, when you upload set of files with a different structure, strings will be auto-translated here.

Files deleted in GitHub will be automatically deleted in Crowdin too, so ideally you won't see obsolete files in your Crowdin project.

By the way, will you use a single branch in Crowdin (e.g. `master`) or multiple branches at a time?

**Comment by HashimotoYT** [comment]
Yes, we use only the `master` branch with all of our content.

So that means we can easily migrate to a new folder structure and have the same localization for other versions without losing the current status?

**Comment by Andrulko** [comment]
> So that means we can easily migrate to a new folder structure and have the same localization for other versions without losing the current status?

I do not see any obstacles here. Once you upload new folder, Crowdin will do the rest (i.e. migrate existing translations). All translations' history will be kept inside new folder too

**Pull request merged**